dm/syncer/dbconn/utils.go

```diff
@@ -96,3 +98,30 @@ func getSessionVariable(tctx *tcontext.Context, conn *DBConn, variable string) (
 	}
 	return value, nil
 }
+
+// TCPConnWithIOCounter is a wrapper of net.TCPConn with counter that accumulates
+// the bytes this connection reads/writes.
+type TCPConnWithIOCounter struct {
+	*net.TCPConn
+	c *atomic.Uint64
```

Comment:
> I'm wondering how we retrieve the network consumption from the TCPConnWithIOCounter?

this is only internally used so short name is tolerated. The caller owns the atomic variable and pass the reference to `NewTCPConnWithIOCounter`, and caller will read atomic variable later.

So engine can `NewAtomicUint64`, set it to config, start a worker and read the value from it later

> how much difference is there between the data size we write to the downstream

This should take effect at the bottom-most layer I can find 🤔

Reply:
> this is only internally used so short name is tolerated. The caller owns the atomic variable and pass the reference to NewTCPConnWithIOCounter, and caller will read atomic variable later.

lgtm

> This should take effect at the bottom-most layer I can find

Maybe we can try on this approach in this version since it seems accessible to collect the stats in this way. If we received any suspicion or complaint from the user, then we should consider a different way.